File: app/src/main/java/de/fhswf/fahrradverleih/android/view/startup/viewmodel/StartupViewModel.java
package de.fhswf.fahrradverleih.android.view.startup.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.viewmodel.CreationExtras;
import androidx.lifecycle.viewmodel.ViewModelInitializer;

import de.fhswf.fahrradverleih.android.FahrradverleihApp;
import de.fhswf.fahrradverleih.android.api.RadApi;
import de.fhswf.fahrradverleih.android.api.remote.RemoteRadApi;

public class StartupViewModel extends ViewModel {

    private final RadApi api;

    private final MutableLiveData<StartupContent> content =
            new MutableLiveData<>(StartupContent.AUTHENTICATION);

    private final MutableLiveData<AuthenticationStatus> authenticationStatus =
            new MutableLiveData<>(AuthenticationStatus.IDLE);

    private final MutableLiveData<String> email = new MutableLiveData<>("");
    private final MutableLiveData<String> password = new MutableLiveData<>("");

    public StartupViewModel(RadApi api) {
        this.api = api;
    }

    public void setContent(@NonNull StartupContent content) {
        this.content.setValue(content);

        switch (content) {
            case AUTHENTICATION:
                authenticate();
                break;
            case LOGIN:
                setAuthenticationStatus(AuthenticationStatus.IDLE);
                setEmail("");
                setPassword("");
                break;
        }
    }

    /**
     * Geht mit Navigationsevents um.
     */
    public void backPressed() {
        switch (getContentValue()) {
            case AUTHENTICATION:
            case WELCOME:
                return;
            case LOGIN:
            case REGISTER:
            default:
                setContent(StartupContent.AUTHENTICATION);
        }
    }

    /**
     * Löst den Login-Prozess aus (erstellt eine neue Session).
     */
    public void login() {
        if (getAuthenticationStatusValue() != AuthenticationStatus.IDLE) return;
        setAuthenticationStatus(AuthenticationStatus.AUTHENTICATING);
        api.login(
                getEmailValue(), getPasswordValue(),
                result -> {
                    if (api instanceof RemoteRadApi)
                        ((RemoteRadApi) api).updateToken(result.getToken());
                    setContent(StartupContent.AUTHENTICATION);
                },
                e -> {
                    setAuthenticationStatus(AuthenticationStatus.FAILED);
                    setAuthenticationStatus(AuthenticationStatus.IDLE);
                }
        );
    }

    /**
     * Prüft, ob die Authentifizierung bei der Rad-API gegeben ist.
     * <p>
     * Falls nicht, wird auf die Willkommensseite gewechselt, auf welcher der
     * Anwender sich (erneut) anmelden kann.
     */
    private void authenticate() {
        setAuthenticationStatus(AuthenticationStatus.AUTHENTICATING);
        api.auth(
                (res) -> setAuthenticationStatus(AuthenticationStatus.AUTHENTICATED),
                (e) -> {
                    setAuthenticationStatus(AuthenticationStatus.FAILED);
                    setAuthenticationStatus(AuthenticationStatus.IDLE);
                    setContent(StartupContent.WELCOME);
                }
        );
    }

    public LiveData<StartupContent> getContent() {
        return content;
    }

    @NonNull
    public StartupContent getContentValue() {
        if (content.getValue() == null) throw new IllegalStateException();
        return content.getValue();
    }

    public LiveData<AuthenticationStatus> getAuthenticationStatus() {
        return authenticationStatus;
    }

    @NonNull
    public AuthenticationStatus getAuthenticationStatusValue() {
        if (authenticationStatus.getValue() == null) throw new IllegalStateException();
        return authenticationStatus.getValue();
    }

    public void setAuthenticationStatus(@NonNull AuthenticationStatus authenticationStatus) {
        this.authenticationStatus.setValue(authenticationStatus);
    }

    public LiveData<String> getEmail() {
        return email;
    }

    @NonNull
    public String getEmailValue() {
        if (email.getValue() == null) throw new IllegalStateException();
        return email.getValue();
    }

    public void setEmail(@NonNull String email) {
        this.email.setValue(email);
    }

    public LiveData<String> getPassword() {
        return password;
    }

    @NonNull
    public String getPasswordValue() {
        if (password.getValue() == null) throw new IllegalStateException();
        return password.getValue();
    }

    public void setPassword(@NonNull String password) {
        this.password.setValue(password);
    }

    /**
     * Eigene ViewModel-Factory/Initializer.
     * <p>
     * Liefert in der Funktion für die Erstellung einer ViewModel-Instanz
     * {@link CreationExtras} mit, durch welche unter anderem die Anwendungsinstanz
     * bezogen werden kann. Diese dient als Context für die {@link RemoteRadApi}.
     */
    public static final ViewModelInitializer<StartupViewModel> initializer =
            new ViewModelInitializer<>(
                    StartupViewModel.class,
                    creationExtras -> {
                        var app = (FahrradverleihApp) creationExtras.get(
                                ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY);
                        if (app == null)
                            throw new IllegalStateException("Application must not be null!");

                        return new StartupViewModel(new RemoteRadApi(app));
                    }
            );
}

File: app/src/main/java/de/fhswf/fahrradverleih/android/view/startup/viewmodel/StartupViewModel.kt
package de.fhswf.fahrradverleih.android.view.startup.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewmodel.CreationExtras
import de.fhswf.fahrradverleih.android.api.RadApi
import de.fhswf.fahrradverleih.android.api.remote.RemoteRadApi

enum class StartupContent {
    AUTHENTICATION,
    LOGIN,
    REGISTER,
    WELCOME
}

enum class AuthenticationStatus {
    IDLE, AUTHENTICATING, FAILED, AUTHENTICATED
}

class StartupViewModel(
    private val api: RadApi,
) : ViewModel() {

    private val content = MutableLiveData(StartupContent.AUTHENTICATION)
    private val authenticationStatus = MutableLiveData(AuthenticationStatus.IDLE)

    private val email = MutableLiveData("")
    private val password = MutableLiveData("")

    fun getContent(): LiveData<StartupContent> = content
    fun getAuthenticationStatus(): LiveData<AuthenticationStatus> = authenticationStatus

    fun setEmail(email: String) {
        this.email.value = email
    }

    fun setPassword(password: String) {
        this.password.value = password
    }

    /**
     * Ändert {@link content} und führt ggf. eine mit dem neuen Inhalt verbundene,
     * anfängliche Aktion aus.
     */
    fun navigateTo(content: StartupContent) {
        this.content.value = content

        when (this.content.value) {
            StartupContent.AUTHENTICATION -> {
                this.authenticationStatus.value = AuthenticationStatus.IDLE
                authenticate()
            }
            StartupContent.LOGIN -> {
                authenticationStatus.value = AuthenticationStatus.IDLE
                email.value = ""
                password.value = ""
            }
            else -> {}
        }
    }

    /**
     * Erstellt eine neue Session mit den vom Anwender bereitgestellten Anmeldedaten.
     */
    fun login() {
        if (this.authenticationStatus.value != AuthenticationStatus.IDLE) return
        this.authenticationStatus.value = AuthenticationStatus.AUTHENTICATING
        api.login(email.value!!, password.value!!,
            { result ->
                if (api is RemoteRadApi) api.updateToken(result.token)
                navigateTo(StartupContent.AUTHENTICATION)
            },
            {
                this.authenticationStatus.value = AuthenticationStatus.FAILED
                this.authenticationStatus.value = AuthenticationStatus.IDLE
            }
        )
    }

    /**
     * Geht mit Navigationsevents um.
     *
     * @return true -> Event an Avtivity weitergeben
     *         false -> ViewModel ist zuständig
     */
    fun backPressed(): Boolean {
        return when (this.content.value) {
            StartupContent.AUTHENTICATION -> true
            StartupContent.WELCOME -> true
            else -> {
                navigateTo(StartupContent.AUTHENTICATION)
                false
            }
        }
    }

    private fun authenticate() {
        this.authenticationStatus.value = AuthenticationStatus.AUTHENTICATING
        api.auth(
            { this.authenticationStatus.value = AuthenticationStatus.AUTHENTICATED },
            {
                this.authenticationStatus.value = AuthenticationStatus.FAILED
                this.authenticationStatus.value = AuthenticationStatus.IDLE
                navigateTo(StartupContent.WELCOME)
            }
        )
    }

    companion object {
        /**
         * Factory erlaubt die Instantiierung des ViewModels mit Extras, welche unter anderem
         * eine Instanz der {@link android.app.Application} enthalten (hier für die Erzeugung
         * einer API-Instanz erforderlich).
         */
        val Factory: ViewModelProvider.Factory = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val application =
                    extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]!!

                return StartupViewModel(
                    api = RemoteRadApi(application)
                ) as T
            }
        }
    }
}
